README
WebService-Muzu

Perl interface to the www.muzu.tv API

API documentation is at: www.muzu.tv/api

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc WebService::Muzu

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=WebService-Muzu

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/WebService-Muzu

    CPAN Ratings
        http://cpanratings.perl.org/d/WebService-Muzu

    Search CPAN
        http://search.cpan.org/dist/WebService-Muzu/


LICENSE AND COPYRIGHT

Copyright (C) 2012 Travis Basevi

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for WebService-Muzu

0.01    2012-10-15
        First version, released on an unsuspecting world.



t/MyLWP.pm
package MyLWP;

use LWP::UserAgent;
use HTTP::Response;
use Test::MockObject::Extends;

sub useragent
	{
	my $ua = LWP::UserAgent->new();
	$ua = Test::MockObject::Extends->new($agent);
	$ua->mock('get', \&response);
	return $ua;
	}

sub response
	{
	my $self = shift;
	my $url = shift;

	my $file = $url;
	$file =~ s!^(http://)?www.muzu.tv/api/!!;
	$file =~ s/\?.*$//;
	$file =~ s!/!_!g;

	if (open(IN, "t/$file.xml"))
		{
		local $/ = undef;
		$content = <IN>;
		close(IN);
		return HTTP::Response->new(200, '', ['Content-Type', 'test/html'], $content);
		}
	else
		{
		return HTTP::Response->new(500, "cannot open t/$file.xml: $!");
		}
	}

1;


lib/WebService/Muzu.pm
package WebService::Muzu;

use 5.006;
use strict;
use warnings;

use WebService::Muzu::Artist;
use WebService::Muzu::Channel;
use WebService::Muzu::Image;
use WebService::Muzu::Video;

use parent 'WebService::Muzu::Base';

=head1 NAME

WebService::Muzu - Perl interface to the www.muzu.tv API

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

Muzu has a developer's API documented at http://www.muzu.tv/api/

This module makes life easier perl-wise, building in the fetching of the
XML and the processing of it

    use WebService::Muzu;

    my $foo = WebService::Muzu->new();
    ...

=head1 CONSTRUCTOR

=head2 new

Create a L<WebService::Muzu> instance

=cut

sub _init
	{
	my $self = shift;
	$self->SUPER::_init(@_);

	#$hash->{soundoff} = "n";
	#$hash->{autostart} = "n";
	$self->{_query}->{videotype} = "hq";
	# width
	# height
	# includeAll

	$self->{url_base} = "http://www.muzu.tv/api";
	}

=head1 METHODS

=head2 get_results_by_browse

=cut

sub get_results_by_browse
	{
	my $self = shift;
	my $hr = shift || {};
	my %hash = %{$hr};
	$hash{_path} = "browse";
	my $doc = $self->fetch_doc(\%hash);

	$self->{_result} = {};
	$self->{_result}->{videos} = [];
	foreach my $video ($doc->findnodes('./videos/video'))
		{
		my $v = WebService::Muzu::Video->new();
		$v->load_node($video);
		push(@{$self->{_result}->{videos}}, $v);
		}

	return $self;
	}

=head2 get_results_by_search

=cut

sub get_results_by_search
	{
	my $self = shift;
	my $hr = shift || {};
	my %hash = %{$hr};
	$hash{_path} = "search";
	my $doc = $self->fetch_doc(\%hash);

	$self->{_result} = {};
	$self->{_result}->{videos} = [];
	foreach my $video ($doc->findnodes('./videos/video'))
		{
		my $v = WebService::Muzu::Video->new();
		$v->load_node($video);
		push(@{$self->{_result}->{videos}}, $v);
		}

	return $self;
	}

=head2 videos

=cut

sub videos
	{
	my $self = shift;
	return @{$self->{_result}->{videos}};
	}

=head1 AUTHOR

Travis Basevi, C<< <travis at verymetalnoise.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-webservice-muzu at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WebService-Muzu>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WebService::Muzu

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WebService-Muzu>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WebService-Muzu>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WebService-Muzu>

=item * Search CPAN

L<http://search.cpan.org/dist/WebService-Muzu/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2012 Travis Basevi.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of WebService::Muzu


lib/WebService/Muzu/Base.pm
package WebService::Muzu::Base;

use 5.006;
use strict;
use warnings;

use Carp;
use LWP::UserAgent;
use XML::LibXML;

use utf8;
use Encode;

our $AUTOLOAD;

=head1 NAME

WebService::Muzu::Base - core module for the Muzu webservice

=head1 SYNOPSIS

This module shouldn't be called directly, please see
L<../Muzu.pm> for full information

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

sub new
	{
	#my $proto = shift;
	#my $class = ref($proto) || $proto;
	
	my $class = shift;
	my $self = {};
	bless($self, $class);
	$self->_init(@_);

	return $self;
	}

sub _init
	{
	my $self = shift;
	my %config = @_;

	$self->{_result} = {};

	$self->{_query} = {};
	foreach my $k (keys %config)
		{
		if ($k eq 'useragent')
			{
			$self->{useragent} = $config{$k};
			}
		else
			{
			$self->{_query}->{$k} = $config{$k};
			}
		}

	$self->{useragent} ||= LWP::UserAgent->new; # to mock for testing
	$self->{_query}->{format} ||= "xml"; # rss is the other option
	}

sub get_names
	{
	my $self = shift;

	my @names = ();
	foreach my $key (sort keys %{$self->{_result}})
		{
		if (!ref($self->{_result}->{$key}))
			{
			push(@names, $key);
			}
		}

	return @names;
	}

sub get_value
	{
	my $self = shift;
	my $name = shift;

	if (exists $self->{_result}->{$name} && !ref($self->{_result}->{$name}))
		{
		return $self->{_result}->{$name};
		}
	else
		{
		carp "no name '$name' found in $self";
		}
	}

sub AUTOLOAD
	{
	my $self = shift;

	my $name = $AUTOLOAD;
	$name =~ s/.*:://;

	if (exists $self->{_result} && exists $self->{_result}->{$name} && !ref($self->{_result}->{$name}))
		{
		return $self->{_result}->{$name};
		}
	else
		{
		carp "no attribute '$name' found in $self";
		}
	}

sub fetch_doc
	{
	my $self = shift;
	my $query_tmp = shift;
	
	my %query = %{$self->{_query}};
	if ($query_tmp)
		{
		@{\%query}{keys %{$query_tmp}} = values %{$query_tmp};
		#foreach my $k (keys %{$query_tmp})
		#	{
		#	$query{$k} = $query_tmp->{$k};
		#	}
		}

	my $u = $self->{url_base};
	my $qs = "";
	foreach my $k (sort keys %query)
		{
		if ($k eq "_path")
			{
			$u .= "/" . $query{$k};
			}
		elsif ($k !~ /^_/)
			{
			$qs .= $qs ? "&" : "";
			$qs .= _urlencode($k) . "=" . _urlencode($query{$k});
			}
		}
	if ($qs)
		{
		$u .= "?" . $qs;
		}
	$self->{_last_url} = $u;

	my $ua = $self->{useragent};
	my $response = $ua->get($u);
	if ($response->is_success)
		{
		my $content = $response->decoded_content;

#		if (!utf8::is_utf8($content)) # since Perl 5.8.1
#			{
#			$content = decode('iso-8859-1', $content);
#			}

		if ($content)
			{
			my $doc = XML::LibXML->load_xml(string => $content);
			if (my ($error) = $doc->findnodes('/error'))
				{
				die "Muzu error ", $error->getAttribute('code'), ": ", $error->getAttribute('description');
				}
			else
				{
				return $doc;
				}
			}
		else
			{
			croak "no content found for url: $u";
			}
		}
	else
		{
		croak "cannot get url $u: ", $response->status_line;
		}
	}

sub _urlencode
	{
	my $toencode = shift;
	if (defined($toencode))
		{
		utf8::encode($toencode);
		$toencode =~ s/([^A-Za-z0-9\-_.~])/uc sprintf("%%%02X",ord($1))/eg;
		$toencode =~ s/\%20/\+/g;

		return $toencode;
		}
	else
		{
		return undef;
		}
	}

sub DESTROY
	{
	}

=head2 function2

=cut

sub function2 {
}

=head1 AUTHOR

Travis Basevi, C<< <travis at verymetalnoise.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-webservice-muzu at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WebService-Muzu>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 LICENSE AND COPYRIGHT

Copyright 2012 Travis Basevi.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1; # End of WebService::Muzu


lib/WebService/Muzu/Image.pm
package WebService::Muzu::Image;

use 5.006;
use strict;
use warnings;
use Carp;

use parent 'WebService::Muzu::Base';

=head1 NAME

WebService::Muzu - The great new WebService::Muzu!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use WebService::Muzu;

    my $foo = WebService::Muzu->new();
    ...

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

sub _init
	{
	my $self = shift;
	$self->SUPER::_init(@_);

	#$hash->{soundoff} = "n";
	#$hash->{autostart} = "n";
	$self->{_query}->{videotype} = "hq";
	# width
	# height
	# includeAll

	$self->{url_base} = "http://www.muzu.tv/api/image";
	}

=head2 aname

Sets the artist name to be used in the search

=cut

sub get_results_by_video_id
	{
	my $self = shift;
	my $id = shift;

	my $doc = $self->fetch_doc({id => $id, _path => "video"});

	$self->{_result} = {};
	$self->load_node($doc->findnodes('/image'));

	return $self;
	}

sub get_results_by_channel_token
	{
	my $self = shift;
	my $id = shift;

	my $doc = $self->fetch_doc({id => $id, _path => "channel"});

	$self->{_result} = {};
	$self->load_node($doc->findnodes('/image'));

	return $self;
	}

sub image
	{
	return shift;
	}

sub load_node
	{
	my $self = shift;
	my $image = shift;

	foreach my $a ($image->attributes())
		{
		my $name = $a->localName;
		my $value = $a->getValue;
		$self->{_result}->{$name} = $value;
		}

	foreach my $node ($image->getChildrenByTagName('*'))
		{
		my $name = $node->localName;

		if ($name eq 'url')
			{
			my $value = $node->to_literal;
			if (defined $value)
				{
				$value =~ s/^\s+|\s+$//g;
				}
			$self->{_result}->{$name} = $value;
			}
		else
			{
			carp "unknown image childnode tag: $name";
			}
		}
	}

=head1 AUTHOR

Travis Basevi, C<< <travis at verymetalnoise.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-webservice-muzu at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WebService-Muzu>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 LICENSE AND COPYRIGHT

Copyright 2012 Travis Basevi.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;


lib/WebService/Muzu/Artist.pm
package WebService::Muzu::Artist;

use 5.006;
use strict;
use warnings;
use Carp;

use parent 'WebService::Muzu::Base';

=head1 NAME

WebService::Muzu - The great new WebService::Muzu!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use WebService::Muzu;

    my $foo = WebService::Muzu->new();
    ...

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

sub _init
	{
	my $self = shift;
	$self->SUPER::_init(@_);

	#$hash->{soundoff} = "n";
	#$hash->{autostart} = "n";
	$self->{_query}->{videotype} = "hq";
	# width
	# height
	# includeAll

	$self->{url_base} = "http://www.muzu.tv/api/artist/details";
	}

=head2 aname

Sets the artist name to be used in the search

=cut

sub get_results_by_name
	{
	my $self = shift;
	my $aname = shift;

	my $doc = $self->fetch_doc({aname => $aname});

	$self->{_result} = {};
	$self->{_result}->{channel} = [];
	foreach my $channel ($doc->findnodes('/artist/channel'))
		{
		my $c = WebService::Muzu::Channel->new();
		$c->load_node($channel);
		push(@{$self->{_result}->{channel}}, $c);
		}

	return $self;
	}

sub channels
	{
	my $self = shift;
	return @{$self->{_result}->{channel}};
	}

sub videos
	{
	my $self = shift;
	my @videos = ();
	foreach my $channel (@{$self->{_result}->{channel}})
		{
		my @v = $channel->videos;
		push(@videos, @v);
		}
	return @videos;
	}

=head1 AUTHOR

Travis Basevi, C<< <travis at verymetalnoise.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-webservice-muzu at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WebService-Muzu>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 LICENSE AND COPYRIGHT

Copyright 2012 Travis Basevi.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;


lib/WebService/Muzu/Channel.pm
package WebService::Muzu::Channel;

use 5.006;
use strict;
use warnings;
use Carp;

use parent 'WebService::Muzu::Base';

=head1 NAME

WebService::Muzu - The great new WebService::Muzu!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use WebService::Muzu;

    my $foo = WebService::Muzu->new();
    ...

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

sub _init
	{
	my $self = shift;
	$self->SUPER::_init(@_);

	$self->{url_base} = "http://www.muzu.tv/api/channel/details";

	#$hash->{soundoff} = "n";
	#$hash->{autostart} = "n";
	$self->{_query}->{videotype} = "hq";
	# width
	# height
	# includeAll
	}

=head2 aname

Sets the artist name to be used in the search

=cut

sub get_results_by_vanityname
	{
	my $self = shift;
	my $vname = shift;

	my $doc = $self->fetch_doc({vname => $vname});

	$self->{_result} = {};
	$self->load_node($doc->findnodes('/channel'));

	return $self;
	}

sub get_results_by_token
	{
	my $self = shift;
	my $token = shift;

	my $doc = $self->fetch_doc({token => $token});

	$self->{_result} = {};
	$self->load_node($doc->findnodes('/channel'));

	return $self;
	}

sub channel
	{
	return shift;
	}

sub image
	{
	my $self = shift;
	return $self->{_result}->{image};
	}

sub thumbnails
	{
	my $self = shift;
	return @{$self->{_result}->{thumbnails}};
	}

sub videos
	{
	my $self = shift;
	return @{$self->{_result}->{videos}};
	}

sub load_node
	{
	my $self = shift;
	my $channel = shift;

	foreach my $a ($channel->attributes())
		{
		my $name = $a->localName;
		my $value = $a->getValue;
		$self->{_result}->{$name} = $value;
		}

	foreach my $node ($channel->getChildrenByTagName('*'))
		{
		my $name = $node->localName;

		if ($name eq 'name' || $name eq 'description' || $name eq 'tags' || $name eq 'url' || $name eq 'embed')
			{
			my $value = $node->to_literal;
			if (defined $value)
				{
				$value =~ s/^\s+|\s+$//g;
				}
			$self->{_result}->{$name} = $value;
			}
		elsif ($name eq 'vanity')
			{
			next; # ignore this, only comes from video.channel, and is duplicated data of the vanityname attribute
			}
		elsif ($name eq 'image')
			{
			my $i = WebService::Muzu::Image->new();
			$i->load_node($channel->findnodes('./image'));
			$self->{_result}->{image} = $i;
			}
		elsif ($name eq 'thumbnails')
			{
			$self->{_result}->{thumbnails} = [];
			foreach my $image ($channel->findnodes('./thumbnails/image'))
				{
				my $i = WebService::Muzu::Image->new();
				$i->load_node($image);
				push(@{$self->{_result}->{thumbnails}}, $i);
				}
			}
		elsif ($name eq 'videos')
			{
			$self->{_result}->{videos} = [];
			if ($channel->findnodes('./videos/playlist'))
				{
				foreach my $video ($channel->findnodes('./videos/playlist/video')) # ignoring playlist node (which has a playlist id) as it seems insignificant (can add methods for it later if it becomes meaningful)
					{
					my $v = WebService::Muzu::Video->new();
					$v->load_node($video);
					push(@{$self->{_result}->{videos}}, $v);
					}
				}
			elsif ($channel->findnodes('./videos/video'))
				{
				foreach my $video ($channel->findnodes('./videos/video')) # this logic is for when processing channel via Muzu::Artist
					{
					my $v = WebService::Muzu::Video->new();
					$v->load_node($video);
					push(@{$self->{_result}->{videos}}, $v);
					}
				}
			}
		else
			{
			carp "unknown channel childnode tag: $name";
			}
		}
	}

=head1 AUTHOR

Travis Basevi, C<< <travis at verymetalnoise.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-webservice-muzu at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WebService-Muzu>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 LICENSE AND COPYRIGHT

Copyright 2012 Travis Basevi.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;


lib/WebService/Muzu/Video.pm
package WebService::Muzu::Video;

use 5.006;
use strict;
use warnings;

use Carp;
use XML::LibXML;
use LWP::UserAgent;
use URI ();
use JSON ();

use utf8;
use Encode;

use parent 'WebService::Muzu::Base';

my %month2num = qw(Jan 01 Feb 02 Mar 03 Apr 04 May 05 Jun 06 Jul 07 Aug 08 Sep 09 Oct 10 Nov 11 Dec 12);

=head1 NAME

WebService::Muzu - The great new WebService::Muzu!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use WebService::Muzu;

    my $foo = WebService::Muzu->new();
    ...

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

sub _init
	{
	my $self = shift;
	$self->SUPER::_init(@_);

	#$hash->{soundoff} = "n";
	#$hash->{autostart} = "n";
	$self->{_query}->{videotype} = "hq";
	# width
	# height
	# includeAll

	$self->{url_base} = "http://www.muzu.tv/api/video/details";
	}

=head2 aname

Sets the artist name to be used in the search

=cut

sub get_results_by_video_id
	{
	my $self = shift;
	my $id = shift;

	my $doc = $self->fetch_doc({id => $id});

	$self->{_result} = {};
	$self->load_node($doc->findnodes('/video'));

	return $self;
	}

sub video
	{
	return shift;
	}

sub max_resolution
	{
	my $self = shift;

	my $max_resolution = 0;
	foreach my $k (keys %{$self->{_result}})
		{
		if ($k =~ /^v(\d+)p$/ && $self->{_result}->{$k})
			{
			$max_resolution = $1 > $max_resolution ? $1 : $max_resolution;
			}
		}

	return $max_resolution;
	}

sub download
	{
	my $self = shift;
	my %dl_hash = @_;
	
	if (!$dl_hash{viewhash})
		{
		croak "cannot download without a viewhash argument";
		}

	if ($self->{_result}->{embed})
		{
		my $embed = $self->{_result}->{embed};
		$embed =~ s/\&/&amp;/g;
		$embed =~ s/(<\/object>).+$/$1/sg;

		#use Data::Dumper;
		#print Dumper($embed);

		my $doc = XML::LibXML->load_xml(string => $embed);
		my $url1;
		foreach my $param ($doc->findnodes('/object/param'))
			{
			my %attr = ();
			foreach my $a ($param->attributes())
				{
				$attr{$a->localName} = $a->getValue;
				}
			if ($attr{'name'} eq 'movie')
				{
				$url1 = $attr{'value'};
				last;
				}
			}

		if ($url1)
			{
			#my $ua = $self->{useragent};
			my $ua = LWP::UserAgent->new(max_redirect => 0);
			$ua->timeout(600); # 10 minute timeout
			$ua->default_header('User-Agent' => 'Mozilla/5.0 (Windows NT 6.1; WOW64; rv:16.0) Gecko/20100101 Firefox/16.0'); # appears to ban on libwww-perl
			if ($dl_hash{proxy})
				{
				$ua->proxy('http', $dl_hash{proxy});
				}

			my $r1 = $ua->get($url1);
			if ($r1->code == 302)
				{
				my $url2 = $r1->header('location');
				my $uri2 = URI->new($url2);
				my %qs2 = $uri2->query_form();
				if ($qs2{p})
					{
					my $json2 = JSON->new->utf8(1)->decode($qs2{p});

					my $uri3 = URI->new($json2->{serverURL});
					$uri3->path('/player/playerInit');
					$uri3->query_form([
						pi => 0,
						areaName => $json2->{areaName},
						ai => $json2->{vidId},
						ni => $json2->{t},
						country => $json2->{country},
						hostName => '',
						randomVideo => 'false',
						networkVersion => $json2->{networkVersion},
						jbMode => 'false',
						loadAll => 'false',
						device => $json2->{device},
						]);
					my $url3 = $uri3->as_string;
					# this doesn't supply anything useful (looking for viewhash)
					#my $r3 = $ua->get($url3);
					#if ($r3->is_success)
					#	{
					#	my $json3 = JSON->new->decode($r3->decoded_content(charset => 'UTF-8'));
					#	}

					my $uri4 = URI->new('http://player.muzu.tv/player/requestVideo');
					$uri4->query_form([
						qv => $self->max_resolution,
						ai => $json2->{vidId},
						cn => 0, # count number in playlist?
						tm => 0, # offset in seconds
						networkVersion => '1265919724', # $json2->{networkVersion},
						vt => 'y',
						country => $json2->{country},
						viewhash => $dl_hash{viewhash},
						hostName => 'http://www.muzu.tv',
						areaName => 'channel', # $json2->{areaName},
						playlistId => 0,
						ni => '81467', # $json2->{t},
						device => 'web.onsite', # $json2->{device},
						]);

					my $url4 = $uri4->as_string;

					my $r4 = $ua->get($url4);
					if ($r4->is_success)
						{
						my $json4 = JSON->new->utf8(1)->decode($r4->decoded_content);

						if ($json4->{url} eq "http://player.muzu.tv/player/invalidVideo")
							{
							carp "cannot download video: \"invalidVideo\"";
							return undef;
							}
						elsif ($json4->{url} eq "http://player.muzu.tv/player/invalidDevice")
							{
							carp "cannot download video: \"invalidDevice\"";
							return undef;
							}
						elsif ($json4->{url} eq "http://player.muzu.tv/player/invalidTerritory")
							{
							carp "cannot download video: \"invalidTerritory\"";
							return undef;
							}
						else
							{
							my $uri5 = URI->new($json4->{url});
							my ($ext) = $uri5->path =~ /.+\.(.+)$/;
							#my $filename = "D:/Video/Music/Youtube/" . $json2->{vidId} . "." . $ext;
							my $filename = $json2->{vidId} . "." . $ext;
							my $url5 = $uri5->as_string;
							my $r5 = $ua->get($url5, ':content_file' => $filename);
							if ($r5->is_success)
								{
								#my $size_header = $ua->default_header("Content-Length");
								#my $size_file = (stat($filename))[7];
								#if ($size_header != $size_file)
								#	{
								#	carp sprintf("warning! byte lengths do not match: header %d != file %d", $size_header, $size_file);
								#	}
								#print "file written to $filename\n";
								return $filename;
								}
							elsif ($r5->code == 403)
								{
								carp "cannot download video: 403 Forbidden";
								return undef;
								}
							else
								{
								croak "error downloading $url5: ", $r5->status_line;
								}
							}
						}
					else
						{
						croak "error downloading $url4: ", $r4->status_line;
						}
					}
				else
					{
					croak "unknown query string for url $url1: p not found";
					}
				}
			elsif ($r1->code == 500 && $r1->status_line =~ /Connection refused/)
				{
				carp "cannot download video: \"Connection refused\"";
				return undef;
				}
			else
				{
				croak "302 expected for url $url1: ", $r1->status_line;
				}
			}
		else
			{
			croak "cannot find movie value in embed html: ", $embed;
			}
		}
	else
		{
		croak "no embed html found to parse for download (have you loaded a video into the object?)";
		}
	}

=head2 channel_vanityname

Channel vanityname which can then be passed to ::Channel get_results_by_vanity_name

=cut

sub channel_vanityname
	{
	my $self = shift;
	return $self->{_result}->{channel}->{vanityname};
	}

sub thumbnails
	{
	my $self = shift;
	return @{$self->{_result}->{thumbnails}};
	}

sub load_node
	{
	my $self = shift;
	my $video = shift;

	foreach my $a ($video->attributes())
		{
		my $name = $a->localName;
		my $value = $a->getValue;
		if ($value eq 'true')
			{
			$self->{_result}->{$name} = 1;
			}
		elsif ($value eq 'false')
			{
			$self->{_result}->{$name} = 0;
			}
		elsif ($value =~ /^(\d\d)\-(Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec)\-(\d\d\d\d)$/)
			{
			$self->{_result}->{$name} = $3 . "-" . $month2num{$2} . "-" . $1;
			}
		else
			{
			$self->{_result}->{$name} = $value;
			}
		}

	foreach my $node ($video->getChildrenByTagName('*'))
		{
		my $name = $node->localName;

		if ($name eq 'title' || $name eq 'description' || $name eq 'artistname' || $name eq 'tags' || $name eq 'url' || $name eq 'embed')
			{
			my $value = $node->to_literal;
			if (defined $value)
				{
				$value =~ s/^\s+|\s+$//g;
				}
			$self->{_result}->{$name} = $value;
			}
		elsif ($name eq 'image')
			{
			my $i = WebService::Muzu::Image->new();
			$i->load_node($video->findnodes('./image'));
			$self->{_result}->{image} = $i;
			}
		elsif ($name eq 'thumbnails')
			{
			$self->{_result}->{thumbnails} = [];
			foreach my $image ($video->findnodes('./thumbnails/image'))
				{
				my $i = WebService::Muzu::Image->new();
				$i->load_node($image);
				push(@{$self->{_result}->{thumbnails}}, $i);
				}
			}
		elsif ($name eq 'channel')
			{
			# not a full set of channel data, thus just maintain a hash
			# could allow a $video->channel method which fetches the channel data but would imply duplication of data as it would fetch the same video data
			$self->{_result}->{channel} = {};
			my ($channel) = $video->findnodes('./channel');
			foreach my $a ($channel->attributes())
				{
				$self->{_result}->{channel}->{$a->localName} = $a->getValue;
				}
			foreach my $cnode ($channel->getChildrenByTagName('*'))
				{
				$self->{_result}->{channel}->{$cnode->localName} = $cnode->to_literal;
				$self->{_result}->{channel}->{$cnode->localName} =~ s/^\s+|\s+$//g;
				}
			}
		else
			{
			carp "unknown video childnode tag: $name";
			}
		}
	}

=head1 AUTHOR

Travis Basevi, C<< <travis at verymetalnoise.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-webservice-muzu at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WebService-Muzu>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 LICENSE AND COPYRIGHT

Copyright 2012 Travis Basevi.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

1;
